Create repo labels using API

sandy3w opened the pull request:
- Addresses #3 

file overview:

AI Skills Assessor: Repo Labeler
.GitHub/workflows/repo-update.yml
-> stores the actual workflow
-> make sure PAT is stored as a secret named PAT, or change all mentions of secrets.PAT to secrets.<new-name>

API_repo_labels/data
-> run excel_to_json.py to convert excel stored at excel_labels_data.xlsx to labels_data.json
-> replace excel and rerun excel_to_json.py when label data has changed

API_repo_labels/label_delete.py
-> deletes all label from repo specified on line 76 and line 77
-> remember to set PAT using: 
	export GH_TOKEN=your_actual_personal_access_token_here

API_repo_labels/repo.labeler.py
-> uses label data in labels_data.json to label repo

config/target_repos_status.json
-> ensure bot has access to modify these repos
-> follow the below format when adding a repo to label: 
"<repo-owner>/<repo-name>": {
      "labeled": false,
      "date": ""
    }
-> successful files will be edited by workflow
-> ensure each label is separated by comma except last one

sandy3w committed:
completed labeler script, need to test

sandy3w committed:
handles empty description

sandy3w committed:
debugging workflow (printing descriptions)

sandy3w committed:
debugging workflow (print description + delete previous labels)

sandy3w committed:
debuggine workflow (checks entire words in readme instead of snippets)

sandy3w committed:
debugging workflow (split concept by "," and "/" instead of white spaces

sandy3w committed:
debugging workflow (updated keywords list)

sandy3w committed:
debugging workflow (print matches)

sandy3w committed:
debugging workflow (changed up IS keyword)

sandy3w committed:
updated workflow (creates all tags)

sandy3w committed:
PAT test

sandy3w committed:
PAT check 2

sandy3w committed:
repos label debug

sandy3w committed:
repos label debug (changed targeted file path)

sandy3w committed:
repos label debug (download jq)

sandy3w committed:
repos label debug (fixed repetition handling)

sandy3w committed:
repos label debug (sleep)

sandy3w committed:
repos label debug (error handling fixed)

sandy3w committed:
repos label debug (changed up repetition handling)

sandy3w committed:
repos label debug (updating status)

sandy3w committed:
repos label debug (added more time buffer)

sandy3w committed:
repos label debug (debug json file)

Commit:
Update label status for target repos

sandy3w committed:
repos label debug (corner cases test)

sandy3w committed:
repos label debug (repeated label)

Commit:
Update label status for target repos

aisa-bot committed:
test

aisa-bot committed:
test2

aisa-bot committed:
bot commit test works

aisa-bot committed:
test for 213 labels

aisa-bot committed:
repetition label handling updated

aisa-bot committed:
Update target_repos_status.json

Commit:
Update label status for target repos

aisa-bot committed:
Update target_repos_status.json

Commit:
Update label status for target repos

sandy3w committed:
comments added

JasonUranta reviewed (commented):
Dear Sandy,
Overall great job!! I admire the work you did on this issue and even took some notes on your coding practices to use for myself going forward. Below is my review:


- Your Python files and YAML are clear, organized, and the logic is straightforward. 
- I appreciated your overview of what each file you are adding/modifying does in your initial PR. 
- I don't have a strong background in bash scripting so I struggle to fully understand the while loop in your YAML file but from what I could piece together it does look good. 
  - However, I'd suggest someone else looks over it as well to make sure.

- Something I noticed is that you wrapped your API requests in try-catches to account for a failed request. This is great but I'd suggest to keep the practice going in repo_labeler.py around lines 28 and 60. 

- Another suggestion I have is to move your Python file excel_to_json.py out of the data folder and put all the .py files in their own folder called "scripts" or something similar for clarity but this is a simple organizational preference of mine.


From,
JasonUranta

chinaexpert1 commented:
Notes on Sandy’s pull request

YAML Looks good.

Could be ready now to add the dedicated issue for error messages and "no labels needed" messages, so it can post a comment to the issue

One thought is the script immediately exits on the first API request failure (fetching labels). If a network error or intermittent GitHub outage occurs, this stops the entire labeling process abruptly. This is fine for now, but maybe we want it to retry depending on the error?
This is a possible modification with an exponential backoff:

```
import requests.exceptions

def fetch_existing_labels(owner, repo, headers, retries=3, backoff=5):
    page = 1
    existing_labels = set()

    while True:
        for attempt in range(retries):
            try:
                response = requests.get(
                    f"https://api.github.com/repos/{owner}/{repo}/labels",
                    headers=headers,
                    params={"per_page": 100, "page": page}
                )
                response.raise_for_status()
                labels_page = response.json()
                break  # successful fetch, exit retry loop
            except requests.exceptions.RequestException as e:
                print(f"Attempt {attempt + 1}/{retries} failed: {e}")
                time.sleep(backoff * (2 ** attempt))
        else:
            print("Exceeded retry limit fetching labels.")
            sys.exit(1)

        if not labels_page:
            break

        existing_labels.update(label['name'].strip().lower() for label in labels_page)
        page += 1

    return existing_labels
```
And similarly for rate limit management we can retry if we hit the limit, by sleeping:


```
elif create_resp.status_code == 403:
    if create_resp.headers.get("X-RateLimit-Remaining") == "0":
        reset_time = int(create_resp.headers["X-RateLimit-Reset"])
        sleep_duration = max(reset_time - int(time.time()) + 5, 5)
        print(f"Rate limit reached, sleeping for {sleep_duration}s...")
        time.sleep(sleep_duration)
        continue  # retry after sleeping
    else:
        print(f"Access forbidden: {create_resp.text}")
```


Similarly for delete_labels.py:

```
from requests.exceptions import RequestException
import time

def get_all_labels(owner, repo, headers, retries=3, backoff=5):
    labels = []
    page = 1

    while True:
        url = f"https://api.github.com/repos/{owner}/{repo}/labels?per_page=100&page={page}"
        attempt = 0
        while attempt < retries:
            try:
                response = requests.get(url, headers=headers, timeout=10)
                response.raise_for_status()
                batch = response.json()
                break  # exit retry loop on success
            except RequestException as e:
                attempt += 1
                print(f"Request failed ({attempt}/{retries}): {e}")
                if attempt < retries:
                    time.sleep(backoff * attempt)
                else:
                    print("Max retries reached, exiting.")
                    sys.exit(1)
        
        if not batch:
            break
        labels.extend(batch)
        page += 1
    return labels
```


And rate limiting for label deletes:

```
def handle_rate_limit(response):
    if response.status_code == 403 and response.headers.get("X-RateLimit-Remaining") == "0":
        reset_time = int(response.headers["X-RateLimit-Reset"])
        wait_seconds = max(reset_time - int(time.time()) + 5, 5)
        print(f"Rate limit reached, sleeping for {wait_seconds} seconds...")
        time.sleep(wait_seconds)
        return True
    return False

# In delete loop:
if del_response.status_code == 204:
    print(f"Deleted label: {label_name}")
else:
    if handle_rate_limit(del_response):
        continue  # Retry deletion after sleep
    print(f"Failed to delete label: {label_name} - HTTP {del_response.status_code}: {del_response.text}")
```


Currently, all outputs are via print statements, not stored or systematically logged.
We should use Python’s built-in logging module for better tracking and auditing


```
import logging

logging.basicConfig(filename='delete_labels.log', level=logging.INFO,
                    format='%(asctime)s %(levelname)s %(message)s')

# Example:
logging.info(f"Deleted label: {label_name}")
logging.warning(f"Failed to delete label: {label_name}")
```


All this so far is just “nice haves”, but here is an actual error:

In excel_tojson.py you have an incorrect DataFrame used for JSON export

The script builds a filtered and cleaned final_df, but then writes df.to_json(...). So the output JSON may include unwanted columns.
```
final_df = df[['label_name', 'description', 'keywords', 'label_series', 'color']]
json_data = df.to_json(...)  # <--- should be final_df
```
Other than that I think its clear, readable, effective code. I really don’t have many concerns about it.

chinaexpert1 reviewed (approved):
I left my review comments on the issue

chinaexpert1 commented:
As this is our first PR (and my first review) I'd like to ask Bonnie if we are done before merging, looks good to me...

sandy3w commented:
Hiii Jason and Andrew thanks for your reviews! I'll go into my code and make changes as necessary!

sandy3w committed:
Address PR feedback: switch from print to logging, organized file structure, implemented try/catch for API calls

sandy3w committed:
Update target_repos_status.json

sandy3w committed:
updated yaml to account for new paths

sandy3w committed:
Merge branch 'mixin' of https://github.com/sandy3w/ai-skills-assessor into mixin

sandy3w committed:
Update repo_labeler.py

Commit:
Update label status for target repos

sandy3w committed:
Address PR feedback: call retires, rate limit handling

sandy3w committed:
labeler updated

sandy3w committed:
labeler updated 2

Commit:
Update label status for target repos

sandy3w commented:
Hiii all, I went over my code and made the changes, please see files for review again.

Changes include: new file structure, try/catch for fetching labels, exponential backoff, better rate limit handling, changed from print to logging, fixed up final dataframe usage.

Andrew, would you mind clarifying what you mean by: Could be ready now to add the dedicated issue for error messages and "no labels needed" messages, so it can post a comment to the issue

For Andrew again, I didnt end up adding a logging file, instead, all logged statements can be viewed through Github Action, let me know if that works:
<img width="526" height="163" alt="image" src="https://github.com/user-attachments/assets/a970d531-db75-48eb-b148-1ecb7ace88ad" />

chinaexpert1 commented:
> Hiii all, I went over my code and made the changes, please see files for review again.
> 
> Changes include: new file structure, try/catch for fetching labels, exponential backoff, better rate limit handling, changed from print to logging, fixed up final dataframe usage.
> 
> Andrew, would you mind clarifying what you mean by: Could be ready now to add the dedicated issue for error messages and "no labels needed" messages, so it can post a comment to the issue
> 
> For Andrew again, I didnt end up adding a logging file, instead, all logged statements can be viewed through Github Action, let me know if that works:
> <img width="526" height="163" alt="image" src="https://github.com/user-attachments/assets/a970d531-db75-48eb-b148-1ecb7ace88ad" />
> 

Hi Sandy
Great stuff

The dedicated issue hasn't been created yet, it was an idea to have an issue to keep track of error messages from the bot. You can implement this at your leisure next week...it just requires a new issue in the repo called "AISA Bot Error Messages".

Your logging strategy looks fine
Ill approve when I get home later today.

Thanks

Andrew
